text/1003-cli-naming.rst

```diff
@@ -19,110 +19,238 @@ specified.
+complex functionality like the migration subsystem. Better suited avenues
+for this are the online documentation and/or ``man`` pages, or better yet,
+``tldr``.
 
-Overview
-========
+To address discoverability of related sub-commands, top-level listing of
+sub-commands should not be alphabetical but rather grouped by functionality.
 
-The RFC proposes to adopt the ``<group> <command>`` naming scheme for
-the ``edgedb`` CLI.  Here's an outline of current and future ``edgedb``
-subcommands:
+``configure`` uses ``<group> <command>`` whereas ``list-`` is flat
+------------------------------------------------------------------
+
+**Concern**: ``configure`` is inconsistent with ``list-``.
+
+**Decision**: It should be replaced by ``edgedb insert-config``,
+``edgedb set-config``, and``edgedb reset-config``.
+
+Why is ``server`` special?
+--------------------------
+
+**Concern**: the ``edgedb server`` CLI design outlined in RFC 1001
+directly conflicts with the ``<action>-<category>`` scheme.
+
+**Decision**: this sub-system is different because it is in fact
+a tool within a tool. Whereas other commands deal with a single
+database server, often a single database within that server, the
+``edgedb server`` sub-system deals with management of potentially
+many databases and server instances. With that in mind, keeping it
+separate makes sense.
+
+In particular, trying to shoehorn the sub-commands of ``edgedb server``
+into the ``<action>-<object>`` scheme would essentially move all of
+them into flat ``*-instance`` (or ``*-server``) sub-commands.
+
+Are ``list-*`` commands useful?
+-------------------------------
+
+**Concern**: there are nine ``list-*`` sub-commands which is over 36%
+of the available sub-commands. They mirror REPL functionality for listing
+items in the current database, for example ``\la`` or ``\list-aliases``
+will list available aliases. Are all of them necessary?
+
+**Decision**: the CLI should keep listings of entities that it can
+directly manipulate with other sub-commands. This includes:
+
+* ``list-databases`` (through ``create-database``);
+
+* ``list-ports`` (through ``configure``); and
+
+* ``list-roles`` (through ``alter-role``).
+
+Other listings should be removed as they are redundant with REPL
+functionality. Instead, the following form should be enabled::
+
+  $ echo "\list-object-types" | edgedb
+
+This is consistent with behavior of, say, the ``sqlite3`` CLI.
+
+The REPL help for ``\\`` commands is overloaded
+-----------------------------------------------
+
+**Concern**: It is convenient to expose the EdgeDB CLI commands directly in
+REPL. This can greatly simplify administrative tasks when the DB
+administrator has full DB admin rights but yet can't access the server shell.
+Sadly, the presence of those commands under the ``\\`` prefix mixes them
+up with the REPL-specific escape commands like ``\last-error``.
+
+**Decision**: Expose CLI-specific commands in the REPL via the ``!`` prefix,
```

Comment:
(maybe) ACTIONABLE ITEM 7: allow CLI access through a `!` prefix

@1st1  proposed this and I like it. It speaks to my Vim and IPython-trained brain. @tailhook doesn't like it because it will make finding the right option take twice as long. This needs to be decided.

Reply:
I'm biased here, but I still do like the idea even after a few months after proposing it. IMO, repl-specific commands like `\lt` are different from general-purpose shell commands like `dump`.

Reply:
Even if they are different, we will have to do:
```
some-db> \dump my.dump
No such command `dump`.
Did you mean `!dump`?
> !list-databases
Do such command `list-databases`
Did you mean `\list-databases`?
````
Which already smells like we have to parse them equally anyway, just create unnecessary hassle for users.

I still don't see what's the problem with unifying them. In this case, this question doesn't even exist.

Reply:
Alright, I'm now convinced it's indeed a bad idea. We'll introduce confusion where its not really necessary. -1 from me too.

Reply:
Moving this to rejected ideas with Paul's rationale.